File: tests/analog_pulse_unittest.cpp
#include <gtest/gtest.h>
#include "TextFilePulseShape.h"
#include "AnalogPulse.h"

#ifndef PULSE_SHAPE_FILE_PATH
#define PULSE_SHAPE_FILE_PATH "tests/fixtures/unipolar-pulse-shape.dat"
#endif

using namespace cps;

TEST(AnalogPulse, Construction) {
  const TextFilePulseShape* pulseShape = new TextFilePulseShape(PULSE_SHAPE_FILE_PATH);

  double amplitude = 1022.9;
  double pedestal = 15.3;
  double phase = 0.5;
  double deformationLevel = 0.2;
  double noiseMean = 0.01;
  double noiseStdDev = 0.1;

  AnalogPulse analogPulse(
      pulseShape,
      amplitude,
      pedestal,
      phase,
      deformationLevel,
      noiseMean,
      noiseStdDev
      );

  EXPECT_EQ(analogPulse.GetAmplitude(), amplitude);
  EXPECT_EQ(analogPulse.GetPhase(), phase);
  EXPECT_EQ(analogPulse.GetNoiseMean(), noiseMean);
  EXPECT_EQ(analogPulse.GetNoiseStdDev(), noiseStdDev);
  EXPECT_EQ(analogPulse.GetPedestal(), pedestal);
  EXPECT_EQ(analogPulse.GetDeformationLevel(), deformationLevel);
  EXPECT_EQ(analogPulse.GetPulseShape(), pulseShape);
}

TEST(AnalogPulse, GetSample) {
  const TextFilePulseShape* pulseShape = new TextFilePulseShape(PULSE_SHAPE_FILE_PATH);
  AnalogPulse analogPulse(pulseShape);
  EXPECT_EQ(analogPulse.GetSample(0), pulseShape->GetY(0));
}

TEST(AnalogPulse, Amplitude) {
  const TextFilePulseShape* pulseShape = new TextFilePulseShape(PULSE_SHAPE_FILE_PATH);
  double amplitude = 1023.0;
  AnalogPulse analogPulse(pulseShape, amplitude);
  EXPECT_EQ(analogPulse.GetSample(0), amplitude * pulseShape->GetY(0));
}

TEST(AnalogPulse, Pedestal) {
  const TextFilePulseShape* pulseShape = new TextFilePulseShape(PULSE_SHAPE_FILE_PATH);
  double pedestal = 50;
  AnalogPulse analogPulse(pulseShape, 1, pedestal);
  EXPECT_EQ(analogPulse.GetSample(-75.5), pedestal + pulseShape->GetY(-75.5));
}

TEST(AnalogPulse, Phase) {
  const TextFilePulseShape* pulseShape = new TextFilePulseShape(PULSE_SHAPE_FILE_PATH);
  AnalogPulse analogPulse;

  double positivePhase = 1.0;
  analogPulse = AnalogPulse(pulseShape, 1, 0, positivePhase);
  EXPECT_EQ(analogPulse.GetSample(0), pulseShape->GetY(1.0));

  double negativePhase = -1.0;
  analogPulse = AnalogPulse(pulseShape, 1, 0, negativePhase);
  EXPECT_EQ(analogPulse.GetSample(0), pulseShape->GetY(-1.0));
}

TEST(AnalogPulse, DeformationLevel) {
  const TextFilePulseShape* pulseShape = new TextFilePulseShape(PULSE_SHAPE_FILE_PATH);
  AnalogPulse analogPulse;
  double deformationLevel;

  deformationLevel = 0.0;
  analogPulse = AnalogPulse(pulseShape, 1, 0, 0, deformationLevel);
  EXPECT_EQ(analogPulse.GetSample(0), pulseShape->GetY(0));

  deformationLevel = 0.5;
  analogPulse = AnalogPulse(pulseShape, 1, 0, 0, deformationLevel);
  EXPECT_NE(analogPulse.GetSample(0), pulseShape->GetY(0));
  double tolerance = deformationLevel * pulseShape->GetYSeries()[151];
  EXPECT_NEAR(analogPulse.GetSample(0), pulseShape->GetY(0), 4 * tolerance);
}

TEST(AnalogPulse, Noise) {
  const TextFilePulseShape* pulseShape = new TextFilePulseShape(PULSE_SHAPE_FILE_PATH);
  AnalogPulse analogPulse;
  double noiseMean, noiseStdDev;

  noiseMean = 0.0;
  noiseStdDev = 0.0;
  analogPulse = AnalogPulse (pulseShape, 1, 0, 0, 0, noiseMean, noiseStdDev);
  EXPECT_EQ(analogPulse.GetSample(0), pulseShape->GetY(0));

  noiseMean = 10.0;
  noiseStdDev = 0.0;
  analogPulse = AnalogPulse(pulseShape, 1, 0, 0, 0, noiseMean, noiseStdDev);
  EXPECT_EQ(analogPulse.GetSample(0), noiseMean + pulseShape->GetY(0));

  noiseMean = 10.0;
  noiseStdDev = 1.0;
  analogPulse = AnalogPulse(pulseShape, 1, 0, 0, 0, noiseMean, noiseStdDev);
  EXPECT_NEAR(analogPulse.GetSample(0), noiseMean + pulseShape->GetY(0), 4 * noiseStdDev);
}

File: python/tests/analog_pulse_unittest.py
# @file python/tests/analog_pulse_unittest.py
# @author Guilherme Inacio Goncalves <ggoncalves@iprj.uer.br>
# @copyright Copyright 2022, Guilherme Inacio Goncalves. All rights reserved.
#
# This file is part of CPS (Calorimetry Pulse Simulator).
#
# CPS is free software: you can redistribute it and/or modify it under the
# terms of the GNU General Public License as published by the Free Software
# Foundation, either version 3 of the License, or any later version.
#
# CPS is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
# details.
#
# You should have received a copy of the GNU General Public License along with
# CPS. If not, see <https://www.gnu.org/licenses/>.
import sys
import os
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '../../build/python'))

import unittest
import pycps as lib

PULSE_SHAPE_FILE_PATH = "tests/fixtures/unipolar-pulse-shape.dat"

class AnalogPulse(unittest.TestCase):

    def test_construction(self):
        pulse_shape = lib.TextFilePulseShape(PULSE_SHAPE_FILE_PATH)
        amplitude = 1022.9
        pedestal = 15.3
        phase = 0.5
        deformation_level = 0.2
        noise_mean = 0.01
        noise_std_dev = 0.1
        analog_pulse = lib.AnalogPulse(
                pulse_shape,
                amplitude,
                pedestal,
                phase,
                deformation_level,
                noise_mean,
                noise_std_dev
                )

        self.assertEqual(analog_pulse.get_amplitude(), amplitude)
        self.assertEqual(analog_pulse.get_pedestal(), pedestal)
        self.assertEqual(analog_pulse.get_phase(), phase)
        self.assertEqual(analog_pulse.get_deformation_level(), deformation_level)
        self.assertEqual(analog_pulse.get_noise_mean(), noise_mean)
        self.assertEqual(analog_pulse.get_noise_std_dev(), noise_std_dev)

    def test_get_sample(self):
        pulse_shape = lib.TextFilePulseShape(PULSE_SHAPE_FILE_PATH)
        amplitude = 100
        pedestal = 50
        analog_pulse = lib.AnalogPulse(pulse_shape, amplitude, pedestal)
        self.assertEqual(
            analog_pulse.get_sample(0),
            pedestal + amplitude * pulse_shape.get_y(0)
        )

if __name__ == '__main__':
    unittest.main()
